locked-asset/proxy_dex/src/wrapped_lp_attributes.rs

```diff
@@ -1,12 +1,117 @@
+}
+
+impl<M: ManagedTypeApi> FixedSupplyToken<M> for WrappedLpTokenAttributes<M> {
+    fn get_total_supply(&self) -> &BigUint<M> {
+        &self.lp_token_amount
+    }
+
+    fn into_part(self, payment_amount: &BigUint<M>) -> Self {
+        let new_lp_token_amount = payment_amount.clone();
+        let new_locked_tokens_amount =
+            self.rule_of_three_non_zero_result(payment_amount, &self.locked_tokens.amount);
+        let new_locked_tokens = EsdtTokenPayment::new(
+            self.locked_tokens.token_identifier,
+            self.locked_tokens.token_nonce,
+            new_locked_tokens_amount,
+        );
+
+        WrappedLpTokenAttributes {
+            lp_token_id: self.lp_token_id,
+            lp_token_amount: new_lp_token_amount,
+            locked_tokens: new_locked_tokens,
+        }
+    }
+}
+
+impl<M: ManagedTypeApi> Mergeable<M> for WrappedLpTokenAttributes<M> {
+    /// locked token nonce can differ, since they get merged through another call
+    fn can_merge_with(&self, other: &Self) -> bool {
+        let same_lp_token = self.lp_token_id == other.lp_token_id;
+        let same_locked_token_id =
+            self.locked_tokens.token_identifier == other.locked_tokens.token_identifier;
+
+        same_lp_token && same_locked_token_id
+    }
+
+    fn merge_with(&mut self, other: Self) {
+        self.error_if_not_mergeable(&other);
+
+        self.lp_token_amount += other.lp_token_amount;
+        self.locked_tokens.amount += other.locked_tokens.amount;
+    }
+}
+
+#[derive(ManagedVecItem, Clone)]
+pub struct WrappedLpToken<M: ManagedTypeApi> {
+    pub payment: EsdtTokenPayment<M>,
+    pub attributes: WrappedLpTokenAttributes<M>,
+}
+
+impl<M: ManagedTypeApi + StorageMapperApi + CallTypeApi> WrappedLpToken<M> {
+    pub fn new_from_payments(
+        payments: &PaymentsVec<M>,
+        wrapped_token_mapper: &NonFungibleTokenMapper<M>,
+    ) -> ManagedVec<M, Self> {
+        wrapped_token_mapper.require_all_same_token(&payments);
+
+        let mut output = ManagedVec::new();
+        for payment in payments {
+            let attributes: WrappedLpTokenAttributes<M> =
+                wrapped_token_mapper.get_token_attributes(payment.token_nonce);
+            let wrapped_lp_token = WrappedLpToken {
+                payment,
+                attributes,
+            };
+
+            output.push(wrapped_lp_token);
+        }
+
+        output
+    }
+}
+
+/// Merges all tokens under a single one, by also merging the locked tokens.
+/// The caller function should handle the minting of the new wrapped LP tokens.
+pub fn merge_wrapped_lp_tokens_through_factory<M: CallTypeApi>(
+    factory_address: ManagedAddress<M>,
+    mut wrapped_lp_tokens: ManagedVec<M, WrappedLpToken<M>>,
+) -> WrappedLpTokenAttributes<M> {
+    let first_item = wrapped_lp_tokens.get(0);
+    wrapped_lp_tokens.remove(0);
+
+    let mut merged_wrapped_lp_attributes =
+        first_item.attributes.into_part(&first_item.payment.amount);
```

Comment:
Should there be a check for lp token_identifier? In `into_part` function you don't check for that.